docs/src/main/asciidoc/security-customization.adoc

```diff
@@ -620,6 +620,62 @@ are using an executor that is capable of propagating the identity (e.g. no `Comp
 to make sure that Quarkus can propagate it. For more information see the
 xref:context-propagation.adoc[Context Propagation Guide].
 
+[[observe-security-events]]
+== Observe security events
+
+Quarkus beans fire various security events that can be consumed by xref:cdi.adoc#events-and-observers[CDI observers].
+All security events implement the `io.quarkus.security.spi.runtime.SecurityEvent` interface, which allows you to consume all the events with one observer method.
+
+.List of supported security events
+
+* `io.quarkus.security.spi.runtime.AuthenticationFailureEvent`
+* `io.quarkus.security.spi.runtime.AuthenticationSuccessEvent`
+* `io.quarkus.security.spi.runtime.AuthorizationFailureEvent`
+* `io.quarkus.security.spi.runtime.AuthorizationSuccessEvent`
+* `io.quarkus.oidc.SecurityEvent`
+
+[[TIP]]
+For more information about security events specific to the Quarkus OpenID Connect extension, please see
+the xref:security-oidc-code-flow-authentication.adoc#listen-to-authentication-events[Listening to important authentication events]
+section of the OIDC code flow mechanism for protecting web applications guide.
+
+[source,java]
+----
+package org.acme.security;
+
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.spi.runtime.AuthenticationFailureEvent;
+import io.quarkus.security.spi.runtime.AuthorizationFailureEvent;
+import io.quarkus.security.spi.runtime.SecurityEvent;
+import io.vertx.ext.web.RoutingContext;
+import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.event.ObservesAsync;
+
+public class SecurityEventObserver {
+
+    void observe(@ObservesAsync SecurityEvent event) {              <1>
+        SecurityIdentity identity = event.getSecurityIdentity();
+    }
+
+    void observe(@Observes AuthenticationFailureEvent event) {
+        RoutingContext routingContext = (RoutingContext) event      <2>
```

Comment:
I'd propose just to log something here as suggested in another comment, and instead have a list of event properties which will also be available for all event types

Reply:
Makes sense.

Reply:
Sorry, I actually ended up doing something similar again. Feel free to fight back, but I realized:

- we should show that authenticationfailureevent as always accessible Routingcontext (when it is HTTP based app, which is logical enough), it is important as it gives you access to path, credentials, cookies, headers etc.
- I think it doesn't make any sense to log exception without context and routingcontext is useful.

I swear I wrote this example from scratch only to end up on something like this. Have a look.

Reply:
@michalvavrik The new example is fine, RoutingContext won't work for GRPC I guess if you'll make it work later, but it can clarified in due time as well